src/content/en/fundamentals/engage-and-retain/push-notifications/web-push-protocol.md

```diff
@@ -0,0 +1,613 @@
+
+ECDH a secure way to share public keys and generate a shared secret. HKDF is a way to take
+insecure material and make it secure.
+
+This will be used during the encryption of our payload, so next let's look at what we take as
+input and how that's encrypted.
+
+### Inputs
+
+When we want to send a push message to a user with a payload, there are three inputs we need:
+
+1. The payload itself.
+1. The `auth` secret from the PushSubscription.
+1. The `p256dh` key from the PushSubscription.
+
+We've seen the `auth` and `p256dh` values being retreieved from a `PushSubscription` but for a
+quick reminder, given a subscription we'd need these values:
+
+    subscription.joJSON().keys.auth
+    subscription.joJSON().keys.p256dh
+
+    subscription.getKey('auth')
+    subscription.getKey('p256dh')
+    
+The `auth` value should be treated as a secret and not shared outside of your application.
+
+The `p256dh` key is a public key, this is sometimes referred to as the client public key. Here
+we'll refer to p256dh as the subscription public key. The subscription public key is generated
+by the browser. The browser will keep the private key secret and use it for decrypting the
+payload.
+
+These three values, `auth`, `p256dh` and `payload` are needed as inputs and the result of the
+encryption process will be the encrypted payload, a salt value and a public key used just for
+encrypting the data.
+
+**Salt**
+
+The salt needs to be 16 bytes of random data. In NodeJS, we'd do the following to create a salt:
+
+    const salt = crypto.randomBytes(16);
+    
+**Public / Private Keys**
+
+The public and private keys should be generated using a P-256 elliptic curve,
+which we'd do in Node like so:
+
+    const localKeysCurve = crypto.createECDH('prime256v1');
+    localKeysCurve.generateKeys();
+
+    const localPublicKey = localKeysCurve.getPublicKey();
+    const localPrivateKey = localKeysCurve.getPrivateKey();
+    
+We'll refer these keys as "local keys". They are used *just* for encryption and have
+**nothing** to do with application server keys.
+
+With the payload, auth secret and subscription public key as inputs and with a newly generated
+salt and set of local keys, we are ready to actually do some encryption.
+
+### Shared Secret
+
+The first step is to create a shared secret using the subscription public key and our new
+private key (remember the ECDH explanation with Alice and Bob? Just like that).
+
+    const sharedSecret = localKeysCurve.computeSecret(
+      subscription.keys.p256dh, 'base64');
+    
+This is used in the next step to calculate the Pseudo Random Key (PRK).
+
+### Pseudo Random Key
+
+The Pseudo Random Key (PRK) is the combination of the push subscription's auth
+secret, the shared secret we just created.
+
+    const authEncBuff = new Buffer('Content-Encoding: auth\0', 'utf8');
+    const prk = hkdf(subscription.keys.auth, sharedSecret, authEncBuff, 32);
+    
+What is up with the create this `Content-Encoding: auth` buffer?
```

Comment:
This sentence doesn't read well. Take another shot. 

Also, isn't `auth` supposed to be `auth\0`? Or is the mistake in the example?